File: apps/web/app/lib/live-channel-badges.ts
/**
 * Heuristic quality hints from provider channel titles (no structured codec field in v1).
 */
export function inferStreamQualityHints(channelName: string): string[] {
  const n = channelName.toLowerCase();
  if (/\b(4k|uhd|2160p)\b/.test(n)) return ['4K'];
  if (/\b1080p\b|\bfhd\b|full\s*hd/.test(n)) return ['1080p'];
  if (/\b720p\b/.test(n)) return ['720p'];
  if (/\bhd\b/.test(n)) return ['HD'];
  return [];
}

/** Higher = better inferred tier (for sorting). */
export function streamQualityRank(channelName: string): number {
  const hints = inferStreamQualityHints(channelName);
  if (hints.includes('4K')) return 5;
  if (hints.some((h) => h === '1080p')) return 4;
  if (hints.includes('720p')) return 3;
  if (hints.includes('HD')) return 2;
  return 0;
}

File: apps/android-tv/app/src/main/java/com/iptvtavern/androidtv/domain/parser/LiveChannelBadges.kt
package com.iptvtavern.androidtv.domain.parser

/**
 * Quality badge heuristics — Kotlin port of `apps/web/app/lib/live-channel-badges.ts`.
 *
 * Infers stream quality from channel names. Providers often embed
 * resolution hints like "4K", "FHD", "HD" in the channel title.
 */

private val REGEX_4K = Regex("""\b(4k|uhd|2160p)\b""", RegexOption.IGNORE_CASE)
private val REGEX_FHD = Regex("""\b1080p\b|\bfhd\b|full\s*hd""", RegexOption.IGNORE_CASE)
private val REGEX_720P = Regex("""\b720p\b""", RegexOption.IGNORE_CASE)
private val REGEX_HD = Regex("""\bhd\b""", RegexOption.IGNORE_CASE)

/**
 * Returns quality hints inferred from the channel name.
 * Example: "BBC One 4K" → listOf("4K")
 */
fun inferStreamQualityHints(channelName: String): List<String> {
    return when {
        REGEX_4K.containsMatchIn(channelName) -> listOf("4K")
        REGEX_FHD.containsMatchIn(channelName) -> listOf("1080p")
        REGEX_720P.containsMatchIn(channelName) -> listOf("720p")
        REGEX_HD.containsMatchIn(channelName) -> listOf("HD")
        else -> emptyList()
    }
}

/**
 * Numeric quality rank for sorting (higher = better).
 * 5 = 4K, 4 = 1080p, 3 = 720p, 2 = HD, 0 = unknown.
 */
fun streamQualityRank(channelName: String): Int {
    val hints = inferStreamQualityHints(channelName)
    return when {
        "4K" in hints -> 5
        "1080p" in hints -> 4
        "720p" in hints -> 3
        "HD" in hints -> 2
        else -> 0
    }
}
